### PR TITLE
sublime-merge: askpass bug fix

### LINKED ISSUE #88600: sublime-merge cannot push to remote
**Describe the bug**
When using Sublime-Merge, pushing to remote results in an error message: `.git doesn't look like a git repository.`

**To Reproduce**
1. Install both sublime-merge and git.
2. Open a repository.
3. Commit something, so that the local master is ahead of the remote.
4. Attempt a push.

**Expected behavior**
A password prompt appears, and the push continues as normal.

**Additional context**
Some earlier discussion occurred in [this issue](https://github.com/sublimehq/sublime_merge/issues/742) on the Sublime team's own github. We eventually found that the default.nix for sublime-merge seems to build, patch, and then discard a copy of `ssh-askpass`. This may be the source of the issue.

**Notify maintainers**

@zookatron 

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.
Output:
```
these paths will be fetched (0.05 MiB download, 0.28 MiB unpacked):
  /nix/store/hzrq5ccamfkwdkl9gqqsh3ryprzvs044-bash-interactive-4.4-p23-dev
copying path '/nix/store/hzrq5ccamfkwdkl9gqqsh3ryprzvs044-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.41, NixOS, 20.03.1866.a7c70f2e10b (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.4`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```
Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute: sublime-merge
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by zookatron**
Replace ssh-askpass-sublime executable from sublime-merge that is incompatable with the nix package wrapper. Fixes NixOS/nixpkgs#88600

###### Motivation for this change
Sublime merge has it's own askpass program that is used when pushing/pulling from a git repo using an ssh key with a passphrase and no ssh-agent. Currently the askpass program doesn't work (askpass dialog is never shown) due to it's custom functionality requiring it to pass a custom argv[0] argument to the sublime-merge binary which doesn't work because of the wrapper bash script added by the Nix build process to set various environment variables needed for it to run.

###### Things done
I've replaced the askpass program that is distributed with sublime-merge with a simple bash script that has equivalent functionality and works properly with the Nix package.

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/version-management/sublime-merge/common.nix [hunk 1]
```diff
@@ -53,6 +53,16 @@ in let
...
       runHook preInstall
 
+      # We need to replace the ssh-askpass-sublime executable because the default one
+      # will not function properly, in order to work it needs to pass an argv[0] to
+      # the sublime_merge binary, and the built-in version will will try to call the
+      # sublime_merge wrapper script which cannot pass through the original argv[0] to
+      # the sublime_merge binary. Thankfully the ssh-askpass-sublime functionality is
+      # very simple and can be replaced with a simple bash script.
+      rm ssh-askpass-sublime
+      echo -e "#!${stdenv.shell}\nexec -a /ssh-askpass-sublime $out/.sublime_merge-wrapped" > ssh-askpass-sublime
```
Comment (danieldk) [inline pkgs/applications/version-management/sublime-merge/common.nix:63]: You should be able to use `makeWrapper` here (untested):

```suggestion
  makeWrapper $out/.sublime_merge-wrapped $out/ssh-askpass-sublime \
    --argv0 "ssh-askpass-sublime"
```

Comment (zookatron) [reply pkgs/applications/version-management/sublime-merge/common.nix]: Hi @danieldk, thanks for the suggestion. I wasn't aware that makeWrapper had a --argv0 argument, that is useful. The generated ssh-askpass-sublime script is not quite the same as the manually created version as it also causes some other unnecessary command-line arguments to be passed as well due to it adding "$@", but from my testing this doesn't seem to interfere with it working properly so I've updated the branch with this change.

Comment (danieldk) [reply pkgs/applications/version-management/sublime-merge/common.nix]: Awesome, thanks a lot! I was trying to nix-review this PR, but it seems that the hashes incorrect?

```
hash mismatch in fixed-output derivation '/nix/store/w9n393mak5rgsqzm85j46r0c7rwrcpgy-sublime_merge_build_2011_x64.tar.xz':
  wanted: sha256:0r5qqappaiicc4srk08az2vx42m7b6a75yn2ji5pv4w4085hlrzp
  got:    sha256:0hi9glnvkv46hdzp70dglrc3pj9zlrmcfn4bm6r86hiiragibm89
```

It seems that the same hash is used for the dev and non-dev version?

Comment (danieldk) [reply pkgs/applications/version-management/sublime-merge/common.nix]: Could you add another commit to this PR that fixes the hash(es)?

Comment (zookatron) [reply pkgs/applications/version-management/sublime-merge/common.nix]: Huh, good catch, looks like this was introduced in the last version bump to the package, I don't usually use the dev version so I didn't notice. Just pushed up another commit to this branch that fixes this.

**Review by danieldk — CHANGES_REQUESTED** [review]
Final nitpick and then it's good to merge, could you change the commit message for the version bump to:

```
sublime-merge: 2011 -> 2022
```

**Comment by zookatron** [comment]
> Final nitpick and then it's good to merge, could you change the commit message for the version bump to:
> 
> ```
> sublime-merge: 2011 -> 2022
> ```

Absolutely, though is there a preferred way to indicate that it's just the dev version that's updating, not the stable version? Or does it not matter?

**Comment by danieldk** [comment]
> > Final nitpick and then it's good to merge, could you change the commit message for the version bump to:
> > ```
> > sublime-merge: 2011 -> 2022
> > ```
> 
> Absolutely, though is there a preferred way to indicate that it's just the dev version that's updating, not the stable version? Or does it not matter?

Since `sublime-merge-dev` is an attribute:

```
sublime-merge-dev: 2011 -> 2022
```

Sorry for messing up the example message.

**Comment by zookatron** [comment]
> Since `sublime-merge-dev` is an attribute:
> 
> ```
> sublime-merge-dev: 2011 -> 2022
> ```
> 
> Sorry for messing up the example message.

Got it, that makes sense, I've updated the commit message!

**Review by danieldk — APPROVED** [review]
Thanks a lot for making this fix the hash correction!

**Comment by zookatron** [comment]
No problem, thank you for the help and feedback @danieldk !

**Pull request merged**